File: sdks/typescript/src/tokens.ts
/**
 * The token vector — usage split by class, with the enforced invariants (§5.2).
 *
 * Blending the input classes mis-prices the cached slice badly, so the vector is
 * *always* split by class: `inputUncached / cacheRead / cacheWrite5m /
 * cacheWrite1h / output / reasoning`. The 5m and 1h cache writes are DISTINCT
 * fields (never one flat `cacheWrite`). `reasoning` is DERIVED and embedded
 * within `output` (count of `type:"thinking"` blocks), never a separate input.
 *
 * Mirrors `packages/core/src/valuemaxx/core/tokens.py`. The same invariants are
 * enforced: all counts non-negative, `output >= reasoning`, and (in
 * {@link tokenVectorFromProvider}) cache tokens never exceed `total_input`.
 */

/** Per-attempt token usage, split by the six classes (§5.2). */
export interface TokenVector {
  readonly inputUncached: number;
  readonly cacheRead: number;
  readonly cacheWrite5m: number;
  readonly cacheWrite1h: number;
  readonly output: number;
  readonly reasoning: number;
}

/** Thrown when a token vector violates one of the enforced shape invariants. */
export class TokenInvariantError extends Error {
  public override readonly name = "TokenInvariantError";
}

function assertNonNegative(name: string, value: number): void {
  if (!Number.isInteger(value) || value < 0) {
    throw new TokenInvariantError(`${name} must be a non-negative integer, got ${value}`);
  }
}

/**
 * Construct a {@link TokenVector}, enforcing the always-on invariants (1)/(2)/(6):
 * all counts non-negative and `output >= reasoning` (reasoning lives inside output).
 */
export function tokenVector(v: TokenVector): TokenVector {
  assertNonNegative("inputUncached", v.inputUncached);
  assertNonNegative("cacheRead", v.cacheRead);
  assertNonNegative("cacheWrite5m", v.cacheWrite5m);
  assertNonNegative("cacheWrite1h", v.cacheWrite1h);
  assertNonNegative("output", v.output);
  assertNonNegative("reasoning", v.reasoning);
  if (v.reasoning > v.output) {
    throw new TokenInvariantError(
      `reasoning (${v.reasoning}) must not exceed output (${v.output}); ` +
        "reasoning is derived and embedded within output (§5.2)",
    );
  }
  return Object.freeze({
    inputUncached: v.inputUncached,
    cacheRead: v.cacheRead,
    cacheWrite5m: v.cacheWrite5m,
    cacheWrite1h: v.cacheWrite1h,
    output: v.output,
    reasoning: v.reasoning,
  });
}

/** Sum of the four input-side classes (output/reasoning are not input). */
export function totalInput(v: TokenVector): number {
  return v.inputUncached + v.cacheRead + v.cacheWrite5m + v.cacheWrite1h;
}

/**
 * Build a vector from a raw provider usage object, applying the cache guard.
 *
 * Providers report a single `totalInput` count plus the cache-token subsets;
 * `inputUncached` is *derived* as the remainder. Invariant (3)/(5): the cache
 * tokens can never exceed `totalInput` on a real response — a violation means
 * the usage object was mis-parsed, so we reject rather than silently mis-price
 * (which would otherwise surface as a negative uncached remainder).
 */
export function tokenVectorFromProvider(args: {
  totalInput: number;
  cacheRead: number;
  cacheWrite5m: number;
  cacheWrite1h: number;
  output: number;
  reasoning: number;
}): TokenVector {
  const cacheTotal = args.cacheRead + args.cacheWrite5m + args.cacheWrite1h;
  if (cacheTotal > args.totalInput) {
    throw new TokenInvariantError(
      `cache tokens (${cacheTotal}) exceed total_input (${args.totalInput}); ` +
        "provider usage object is inconsistent (§5.2 invariant 3)",
    );
  }
  return tokenVector({
    inputUncached: args.totalInput - cacheTotal,
    cacheRead: args.cacheRead,
    cacheWrite5m: args.cacheWrite5m,
    cacheWrite1h: args.cacheWrite1h,
    output: args.output,
    reasoning: args.reasoning,
  });
}

File: packages/core/src/atm_core/tokens.py
"""The token vector — usage split by class, with the six enforced invariants (§5.2).

Blending the input classes mis-prices the cached slice badly, so the vector is
*always* split by class: ``input_uncached / cache_read / cache_write_5m /
cache_write_1h / output / reasoning``. The 5m and 1h cache writes are DISTINCT
fields (never one flat ``cache_write``). ``reasoning`` is DERIVED and embedded
within ``output`` (count of ``type:"thinking"`` blocks), never a separate input.

The six invariants:
  1. all counts non-negative;
  2. ``output >= reasoning`` (reasoning lives inside output);
  3. cache tokens never exceed ``total_input`` (provider-shape guard);
  4. 5m/1h are distinct fields;
  5. :meth:`TokenVector.from_provider` rejects a violation of (3);
  6. ``reasoning`` is derived/separate, never double-added into the input side.

Invariants (1), (2), (6) are enforced on every construction; (3)/(5) are the
provider-shape guard applied by :meth:`from_provider` (the internal model may
hold reconciled/synthetic shapes that the raw provider object would not).
"""

from __future__ import annotations

from typing import Annotated

from pydantic import Field, model_validator

from atm_core.base import StrictModel

_NonNegInt = Annotated[int, Field(ge=0)]


class TokenVector(StrictModel):
    """Per-attempt token usage, split by the six classes (§5.2)."""

    input_uncached: _NonNegInt
    cache_read: _NonNegInt
    cache_write_5m: _NonNegInt
    cache_write_1h: _NonNegInt
    output: _NonNegInt
    reasoning: _NonNegInt

    @model_validator(mode="after")
    def _reasoning_within_output(self) -> TokenVector:
        """Invariant (2)/(6): reasoning is embedded within output, never beyond it."""
        if self.reasoning > self.output:
            raise ValueError(
                f"reasoning ({self.reasoning}) must not exceed output ({self.output}); "
                "reasoning is derived and embedded within output (§5.2)"
            )
        return self

    @property
    def total_input(self) -> int:
        """Sum of the four input-side classes (output/reasoning are not input)."""
        return self.input_uncached + self.cache_read + self.cache_write_5m + self.cache_write_1h

    @classmethod
    def from_provider(
        cls,
        *,
        total_input: int,
        cache_read: int,
        cache_write_5m: int,
        cache_write_1h: int,
        output: int,
        reasoning: int,
    ) -> TokenVector:
        """Build a vector from a raw provider usage object, applying the cache guard.

        Providers report a single ``total_input`` token count plus the cache-token
        subsets; ``input_uncached`` is *derived* as the remainder. Invariant (3)/(5):
        the cache tokens (read + 5m write + 1h write) can never exceed ``total_input``
        on a real response — a violation means the usage object was mis-parsed, so we
        reject it rather than silently mis-pricing (which would surface as a negative
        uncached remainder otherwise).
        """
        cache_total = cache_read + cache_write_5m + cache_write_1h
        if cache_total > total_input:
            raise ValueError(
                f"cache tokens ({cache_total}) exceed total_input ({total_input}); "
                "provider usage object is inconsistent (§5.2 invariant 3)"
            )
        return cls(
            input_uncached=total_input - cache_total,
            cache_read=cache_read,
            cache_write_5m=cache_write_5m,
            cache_write_1h=cache_write_1h,
            output=output,
            reasoning=reasoning,
        )


__all__ = ["TokenVector"]
